sycl/source/detail/kernel_impl.cpp

```diff
@@ -21,8 +23,47 @@ kernel_impl::kernel_impl(RT::PiKernel Kernel, const context &SyclContext)
                   std::make_shared<program_impl>(SyclContext, Kernel),
                   /*IsCreatedFromSource*/ true) {}
 
+kernel_impl::kernel_impl(RT::PiKernel Kernel, const context &SyclContext,
+                         std::shared_ptr<program_impl> ProgramImpl,
+                         bool IsCreatedFromSource)
+    : MKernel(Kernel), MContext(SyclContext), MProgramImpl(ProgramImpl),
+      MIsCreatedFromSource(IsCreatedFromSource) {
+
+  RT::PiContext Context = nullptr;
+  PI_CALL(RT::piKernelGetInfo(MKernel, CL_KERNEL_CONTEXT, sizeof(Context),
+                              &Context, nullptr));
+  auto ContextImpl = detail::getSyclObjImpl(SyclContext);
+  if (ContextImpl->getHandleRef() != Context)
+    throw cl::sycl::invalid_parameter_error(
+        "Input context must be the same as the context of cl_kernel");
+  PI_CALL(RT::piKernelRetain(MKernel));
+}
+
+kernel_impl::kernel_impl(const context &SyclContext,
+                         std::shared_ptr<program_impl> ProgramImpl)
+    : MContext(SyclContext), MProgramImpl(ProgramImpl) {}
+
+kernel_impl::~kernel_impl() {
+  // TODO catch an exception and put it to list of asynchronous exceptions
+  if (!is_host()) {
+    PI_CALL(RT::piKernelRelease(MKernel));
+  }
+}
+
 program kernel_impl::get_program() const {
-  return createSyclObjFromImpl<program>(ProgramImpl);
+  return createSyclObjFromImpl<program>(MProgramImpl);
```

Comment:
Suggest moving this implementations to the header as it's one liner.

Reply:
@romanovvlad this is impossible at the time. It requires refactoring program class first as program.h includes program_impl.h and it includes kernel_impl.h resulting in a loop.

Reply:
Ok.